ingestor/runner/shutdown/shutdown.go

```diff
@@ -0,0 +1,75 @@
+package runner
+
+import (
+	"context"
+	"net/http"
+	"os"
+
+	"github.com/Azure/adx-mon/ingestor"
+	"github.com/Azure/adx-mon/pkg/logger"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+const (
+	namespace          = "adx-mon"
+	SHUTDOWN_COMPLETED = "shutdown-completed"
+	SHUTDOWN_REQUESTED = "shutdown-requested"
+)
+
+type ShutDownRunner struct {
+	k8sClient  kubernetes.Interface
+	httpServer *http.Server
+	service    ingestor.Interface
+}
+
+func NewShutDownRunner(cli kubernetes.Interface, http *http.Server, svc ingestor.Interface) *ShutDownRunner {
+	return &ShutDownRunner{
+		k8sClient:  cli,
+		httpServer: http,
+		service:    svc,
+	}
+}
+
+func (r *ShutDownRunner) Run(ctx context.Context) error {
+
+	//get ingestor pod in which this runner is running
+	pod, err := r.k8sClient.CoreV1().Pods(namespace).Get(ctx, os.Getenv("HOSTNAME"), metav1.GetOptions{})
+	if err != nil {
+		logger.Errorf("failed to get pod annotations: %v", err)
+		return err
+	}
+
+	//check if shutdown-completed annotation is set
+	if _, ok := pod.Annotations[SHUTDOWN_COMPLETED]; ok {
+		logger.Infof("shutdown already completed on the pod, skipping shutting down")
+		return nil
+	}
+
+	//shutdown the service
+	if _, ok := pod.Annotations[SHUTDOWN_REQUESTED]; ok {
+		logger.Infof("shutting down the service")
+		if err := r.httpServer.Close(); err != nil {
```

Comment:
What happens if a shutdown is request and the pods just exits unexpectedly?  How can we make this SHUTDOWN_REQUESTED phase idempotent and resumable?

Reply:
Annotations are attached to pod object. If the pod exits  unexpectedly but not deleted, the annotations will stay on it. If the same pod starts again, the annotations will still be there and the Runner will pickup the annotations and resume the shutdown.

On the contrary if the pod is recreated, it will not have the annotations. The operator will have to detect if it still wants the new pod object around and then issue another shutdown on it if needed.